Added messagetype 0x14 and 0x15 for file-access.

OliverDDS opened the pull request:
Added 0x14 and 0x15 for File-access. Extended interface with two new functions, to make them useable.

I implemented the **message-type 0x14** (Read-general-reference) and **0x15** (Write-general-reference), which offers a file-like access to up to 10 register-areas. 
To allocate these additional areas, i introduced two new functions, so that the old interface is stable. 
In **modbus_mapping_new_start_address_extend** and  **modbus_mapping_new_extend**, 
an additional parameter **uint16_t nb_file_registers[]** let you determine the size of the file-register-areas ( 0 for not used files )

Two use the messagetypes, two functions **modbus_read_general_reference** and
**modbus_write_general_reference**, have been added, which offer the simple and normally used query with one subrequest. 

I also extended the unit-tests to verify the implementation.

Take this over to mainline, if you like it.

Best regards,
Oliver

OliverDDS commented:
Hmm, strange. I checked with clang-3.5 :
============================================================================
Testsuite summary for libmodbus 3.1.4
============================================================================
# TOTAL: 1
# PASS:  1
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
Where can i get the log from the tests ?

pboettch commented:
Thanks for your PR. Adding 0x14 and 0x15 is a good thing. 

I have two or three remarks:

1. I think you should consider fixing all your indent and white-space-problems and repush as soon as you can. Ie. no tabs for indentation and alignments. Also there are some incoherent white-space after and before operators.

Do not get desperate, there is a clang-format-file which does quite a good job reproducing the original style. (copy it manually from here if you like: https://github.com/pboettch/libmodbus/blob/clang-format/.clang-format)

2. Add documentation. You need to add man-pages for your new functions inside the doc-directory.

3. Most likely the check is failing because your code produces warnings. When compiling your branch with gcc-7 I'm getting (which is due to the incoherent white-space and indent-stuff):

```
Making all in src
  CC       modbus.lo
../../src/modbus.c: In function ‘modbus_mapping_new_extend’:
../../src/modbus.c:2189:5: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
     if (nb_file_register[i])
     ^~
../../src/modbus.c:2191:4: note: ...this statement, but the latter is misleadingly indented as if it is guarded by the ‘if’
    memset(mb_mapping->file_registers[i],0,nb_file_register[i] * sizeof(uint16_t));
    ^~~~~~
../../src/modbus.c: In function ‘modbus_mapping_new_start_address_extend’:
../../src/modbus.c:2220:19: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
                   if (nb_file_registers[i])
                   ^~
../../src/modbus.c:2222:25: note: ...this statement, but the latter is misleadingly indented as if it is guarded by the ‘if’
                         memset(mb_mapping->file_registers[i],0,nb_file_registers[i] * sizeof(uint16_t));
                         ^~~~~~
```

OliverDDS commented:
Thanks for the hints, i fixed the indention and also the warning about the if clause. 

That's the curse, when working on different PCs, with different OS and different Editors :( 

I also added the documentation for the new functions. When writing it, i'm not really happy about the naming "modbus_mapping_new_start_address_extend", but i haven't got a better name for it, and didn't want to change the current function , to stay backward compatible.

mhei commented:
I would appreciate if you would restructure this PR: divide it in smaller changesets, each with a meaningful commit message for itself. This makes it easier to review. I would propose e.g. first patch which adds the new core functionality, one which add the documentation and one or two which adjusts the tests (e.g. factor out the argv-related stuff as this could stand independently from your topic changes).
Please also squash the whitespace fixes into the "topic" changesets.
Please re-use this PR by using `git push -f ...` in your branch.
I also don't like the approach of extending the mapping stuff. Instead we should consider a more general approach by creating something like a Modbus server context, which can hold multiple mappings in parallel and the new file stuff. However, I have for myself not a really good proposal at hand right now...

OliverDDS commented:
I squashed everything into one commit now.
 About splitting it into smaller commits, I not really sure if i got you right, but splitting the documentation and implementation would not help, because that are different files. The main change is in modbus.c and makes only sense in one big blob.
Only the argv changes ( around 15 lines in each unit-test source ),  could be split, but they are only in, because i have a different setup here and don't like to change the harcoded  names everytime :(

pboettch commented:
@OliverDDS Your manually made coding-style changes weren't enough, I opened a PR on your repo which includes the changes I got after running clang-format like this, feel free to squash them.

Basically I ran

```bash
git diff -U0 HEAD^ | clang-format-diff -p1 -i
```
with the `.clang-format`-file in `libmodbus/`.

@mhei Regarding a server implementation, I agree, the mapping-stuff is not optimal. Have you seen my updated pull-request (modbus_reply_callback())? It is ready for review. It's big, though.

Commit:
Added functions 0x14 and 0x15 ( read and write general-reference )

Commit:
Added documentation for new functions 0x14 and 0x15 and new mappings for file-registers

Commit:
Added test for functions 0x14 and 0x15 and mappings of file-registers

Commit:
Improved commandline for tests, devicenames and baudrate of rtu-test can be provided

OliverDDS commented:
Another try:
-There are several commits now for implementation, documentation, tests and improvement of argv
-using clang-format-diff-3.9 with .clang_format to format it correctly.

Hope, this is now, as it should be.

mhei commented on src/modbus.c:
```diff
@@ -1559,6 +1728,140 @@ int modbus_report_slave_id(modbus_t *ctx, int max_dest, uint8_t *dest)
+        rc /= 2; /* rc is in byte, we count in uint16-steps */
+
+#if 0 // Does it make sense to copy back the response to the caller ? It's a
+      // copy of the request
```
I wondered whether the standard enforces to simply mirror the request content back to the caller or if it would be possible to e.g. return the previous data of the file at this position. But after I've read the standard again, I think it's intension really it to reflect the request back. So I think we can just remove this here.

mhei commented on src/modbus.c:
```diff
@@ -1559,6 +1728,140 @@ int modbus_report_slave_id(modbus_t *ctx, int max_dest, uint8_t *dest)
+/* !!! Take care, that dest must have the size of at least read_nb +1  !!! */
+
+int modbus_read_general_reference(modbus_t *ctx, int file_no, int read_addr,
+                                  int read_nb, uint16_t *dest)
```
I would prefer "modbus_read_file_record" as name to stay as much as possible close to this document: http://www.modbus.org/docs/Modbus_Application_Protocol_V1_1b.pdf
(the Modicon naming looks somewhat "antiquated" to me)

mhei commented on src/modbus.c:
```diff
@@ -1559,6 +1728,140 @@ int modbus_report_slave_id(modbus_t *ctx, int max_dest, uint8_t *dest)
+/* This implements only the simple case with one subrequest. More complex can be
+ * created with raw-message. */
+int modbus_write_general_reference(modbus_t *ctx, int file_no, int write_addr,
+                                   int write_nb, const uint16_t *src)
```
Some as above: modbus_write_file_record?

mhei commented on src/modbus-private.h:
```diff
@@ -39,6 +39,10 @@ MODBUS_BEGIN_DECLS
 #define _RESPONSE_TIMEOUT    500000
 #define _BYTE_TIMEOUT        500000
 
+#define SUB_REQUEST_LENGHT 0x07
```
Beside the typo, I would prefer: _SUB_REQ_HDR_LENGTH

mhei commented on src/modbus.c:
```diff
@@ -985,8 +1014,148 @@ int modbus_reply(modbus_t *ctx, const uint8_t *req,
+        if (rsp_length + nb >= MODBUS_MAX_READ_REGISTERS) {
+            rsp_length = response_exception(
+                ctx, &sft, MODBUS_EXCEPTION_ILLEGAL_DATA_VALUE, rsp, TRUE,
+                "Responselenght exceeds telegram-size %d in subrequests %d in "
```
typo + white space missing

mhei commented on src/modbus.h:
```diff
@@ -92,6 +94,7 @@ MODBUS_BEGIN_DECLS
 #define MODBUS_MAX_WR_WRITE_REGISTERS      121
 #define MODBUS_MAX_WR_READ_REGISTERS       125
+#define MODBUS_MAX_REFERENCE_FILES 10
 
```
NAK. This does not belongs to this group here, since it's a implementation specific maximum, not a maximum defined by the standard or derived from it. So it should stand alone...

OliverDDS replied:
The standard says : 
While it is allowed for the File Number to be in the range 1 to 0xFFFF, it should be noted that
interoperability with legacy equipment may be compromised if the File Number is greater than
10 (0x0A).
So imho this fits there quite well and is not purely implementation specific. ( I renamed it to 
MODBUS_MAX_FILE_RECORDS )

OliverDDS replied:
correction renamed it to: MODBUS_MAX_RECORD_FILES .
I also added the comment from the standard to the header, so one may see, where the number came from.

karlp replied:
We don't restrict registers to 10000, we use the full 16bit address space, there's absolutely no reason to imply that the protocol only allows 10 either.  I strongly agree with @mhei's comment here, this is not part of a standard to limit it to 10.

OliverDDS replied:
I guess you got it wrong, the 10 is the number of files, not the number of records in each file. This might be up to 64k. The maximum number of records is controlled by the mapping-function. You can individually allocate your needed number per file. Up to 64k Records * 10 Files.

karlp replied:
stahp.  straight out of app protocol 1.1b. page 33
```
file number: 0x0001 to 0xFFFF
record number: 0x0000 to 0x270F
```

there are then footnotes that "comaptibility with legacy gear may be poor if you actually _use_ file numbers higher than 10"

using file numbers in the "old" "comaptible" range or higher numbers is a _user_ choice.

OliverDDS replied:
I'm still not convinced, but i changed it to 0xffff.

OliverDDS replied:
I have to remove the checks as well due to the limited range of uint16_t , which makes it less safe, when someone reduces the number :(

mhei commented on src/modbus.h:
```diff
@@ -68,6 +68,8 @@ MODBUS_BEGIN_DECLS
 #define MODBUS_FC_REPORT_SLAVE_ID           0x11
+#define MODBUS_FC_READ_GENERAL_REFERENCE 0x14
+#define MODBUS_FC_WRITE_GENERAL_REFERENCE 0x15
 #define MODBUS_FC_MASK_WRITE_REGISTER       0x16
```
Same as above: I would prefer:
MODBUS_FC_READ_FILE_RECORD and MODBUS_FC_WRITE_FILE_RECORD; also looks like a whitespace problem...

mhei reviewed (commented):
I only reviewed the first patch and stopped for now.

Commit:
Changed c++ comments to c-style

Commit:
Changed naming to write+read file record and _sub_req_hdr_length

Commit:
Changed naming of define for maximum record-files to MODBUS_MAX_RECORD_FILES

Commit:
Updated documentation to new function-names and defines

OliverDDS commented:
I changed the code according to the review comments 
- used c-style comments
- changed naming of the functions and the defines towards the current specification
- updated documentation

Commit:
Increased number of files to 0xffff, due to review finding, removed checks due to limited range

LexusAD commented:
Will this pull request merged into master or is 0x14 implemented in another branch? Maybe consider solving the merge conflicts

OliverDDS commented:
Work was picked up here : https://github.com/stephane/libmodbus/pull/449 and here : https://github.com/stephane/libmodbus/pull/548
So no need to have it three times. 
If you like this better, feel free to use or solve the merged conflict ( which is trivial)  and try to get it merged.
